include/deal.II/particles/property_pool.h

```diff
@@ -0,0 +1,103 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__particles_property_pool_h
+#define dealii__particles_property_pool_h
+
+#include <deal.II/base/array_view.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Particles
+{
+  /**
+   * This class manages a memory space in which particles store their
+   * properties. Because this is dynamic memory and often every particle
+   * needs the same amount, it is more efficient to let this be handled by a
+   * central manager that does not need to allocate/deallocate memory every
+   * time a particle is constructed/destroyed.
+   * The current implementation uses simple new/delete allocation for every
+   * block. Additionally, the current implementation
+   * assumes the same number of properties per particle, but of course the
+   * PropertyType could contain a pointer to dynamically allocated memory
+   * with varying sizes per particle (this memory would not be managed by this
+   * class).
+   * Because PropertyPool only returns handles it could be enhanced internally
+   * (e.g. to allow for varying number of properties per handle) without
+   * affecting its interface.
+   */
+  class PropertyPool
+  {
+  public:
+    /**
+     * Typedef for the handle that is returned to the particles, and that
+     * uniquely identifies the slot of memory that is reserved for this
+     * particle.
+     */
+    typedef double *Handle;
+
+    /**
+     * Define a default (invalid) value for handles.
+     */
+    static const Handle invalid_handle;
+
+    /**
+     * Constructor. Stores the number of properties per reserved slot.
+     */
+    PropertyPool (const unsigned int n_properties_per_slot=1);
+
+    /**
+     * Returns a new handle that allows accessing the reserved block
+     * of memory.
+     */
+    Handle allocate_properties_array ();
```

Comment:
Do we want to make this class a bit more abstract? Some of the properties may not be of type `double`, I feel that the properties of particles are hard coded to be a list of `double`s. This looks more like `class ArrayedPropertyPool : public PropertyPool ...`.

Relating this comment to my other comment, may be we want `PropertyPool` class to be more abstract or make `Particle` class inheritable. A more abstract class for particle properties could help users to implement their own versions instead of having to work around this class.

Reply:
I suppose a possible cheap work-around is to have additional `IntegerHandler` to store integers. Then such a pool can be used by the main `Particle` class to implement custom particles, but I think it's more cumbersome as compared to polymorphism.